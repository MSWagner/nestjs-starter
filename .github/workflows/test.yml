name: Test

on: [push, pull_request]

jobs:
  test:
    name: Test
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2

      - name: Build the docker-compose stack
        run: docker-compose -f 'docker-compose-test.yml' up -d
      - name: Check running containers
        run: docker ps -a
      - name: Check logs
        run: docker logs ananas-dark-backend-postgres-test

      - name: Use Node.js
        uses: actions/setup-node@v1
        with:
          node-version: '10.13.0'
      - name: Install dependencies
        run: yarn
      - name: Build App
        run: yarn build
      - name: Run database migrations
        run: yarn db:migrate:test
      - name: Start tests
        run: yarn test