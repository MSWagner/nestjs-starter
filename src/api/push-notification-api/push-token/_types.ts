import { IsString, MinLength } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";

export class PushTokenDto {
    @IsString()
    @MinLength(3)
    @ApiProperty({ minLength: 3 })
    readonly token: string;
}
