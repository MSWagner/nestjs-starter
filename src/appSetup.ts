import compression from "compression";
import helmet from "helmet";
import rateLimit from "express-rate-limit";

import { INestApplication, ValidationPipe, VersioningType } from "@nestjs/common";

export async function setupApp(app: INestApplication): Promise<void> {
    app.setGlobalPrefix("api");
    app.enableVersioning({
        type: VersioningType.URI
    });

    app.use(compression());
    app.use(helmet());
    app.enableCors();
    app.use(
        rateLimit({
            windowMs: 15 * 60 * 1000, // 15 minutes
            max: 100 // limit each IP to 100 requests per windowMs
        })
    );

    app.useGlobalPipes(new ValidationPipe({ whitelist: true, forbidNonWhitelisted: true }));
}
