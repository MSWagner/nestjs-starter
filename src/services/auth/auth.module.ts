import { Module } from "@nestjs/common";

import { AuthService } from "./auth.service";

import { LocalStrategy } from "./strategies/local.strategy";
import { BearerStrategy } from "./strategies/bearer.strategy";

import { UserModule } from "../user/user.module";

import { PassportModule } from "@nestjs/passport";

import { accessTokenProviders } from "./providers/accessTokenProviders";
import { refreshTokenProviders } from "./providers/refreshTokenProviders";

@Module({
    imports: [UserModule, PassportModule],
    providers: [AuthService, LocalStrategy, BearerStrategy, ...accessTokenProviders, ...refreshTokenProviders],
    exports: [AuthService]
})
export class AuthModule {}
