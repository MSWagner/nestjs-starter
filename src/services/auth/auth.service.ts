import * as bcrypt from "bcrypt";
import moment from "moment";
import * as _ from "lodash";

import { Injectable, Inject } from "@nestjs/common";
import { Repository } from "typeorm";

import { UserService } from "../user/user.service";

import { User } from "../../entities/User.entity";
import { AccessToken } from "../../entities/AccessToken.entity";
import { RefreshToken } from "../../entities/RefreshToken.entity";

import { CONFIG } from "../../config";

export interface IToken {
    accessToken: AccessToken;
    refreshToken: RefreshToken;
}

@Injectable()
export class AuthService {
    constructor(
        @Inject(CONFIG.database.defaultAccessTokenRepoName)
        private readonly accessTokenRepository: Repository<AccessToken>,

        @Inject(CONFIG.database.defaultRefreshTokenRepoName)
        private readonly refreshTokenRepository: Repository<RefreshToken>,

        private readonly userService: UserService
    ) {}

    async generateToken(user: User): Promise<{ accessToken: AccessToken; refreshToken: RefreshToken }> {
        const newRefreshToken = new RefreshToken();
        newRefreshToken.validUntil =
            CONFIG.auth.refreshTokenValidityMS > 0
                ? moment().add(CONFIG.auth.refreshTokenValidityMS, "milliseconds").toDate()
                : null;
        newRefreshToken.user = user;

        const accessToken = await this.generateAuthToken(user);
        const refreshToken = await this.refreshTokenRepository.save(newRefreshToken);

        return {
            accessToken,
            refreshToken
        };
    }

    async generateAuthToken(user: User): Promise<AccessToken> {
        const newAccessToken = new AccessToken();
        newAccessToken.validUntil = moment().add(CONFIG.auth.tokenValidity, "milliseconds").toDate();
        newAccessToken.user = user;

        const accessToken = await this.accessTokenRepository.save(newAccessToken);

        return accessToken;
    }

    async validateUser(username: string, password: string): Promise<User> {
        const user = await this.userService.findOne(username);

        if (_.isNil(user)) {
            return null;
        }

        if (!user.isActive) {
            return null;
        }

        const isValid = await bcrypt.compare(password, user.passwordHash);

        return isValid ? user : null;
    }

    async validateToken(token: string): Promise<User> {
        const accessToken = await this.accessTokenRepository.findOne(token, { relations: ["user"] });

        if (_.isNil(accessToken)) {
            return null;
        }

        const isValid = moment().isSameOrBefore(accessToken.validUntil);

        if (!isValid) {
            await this.accessTokenRepository.delete(accessToken);
        }

        if (!accessToken.user.isActive) {
            return null;
        }

        return isValid ? accessToken.user : null;
    }

    async register(username: string, password: string): Promise<User> {
        return this.userService.createUser(username, password);
    }

    async login(user: User): Promise<IToken> {
        return this.generateToken(user);
    }

    async refreshAuthToken(token: string): Promise<IToken> {
        const refreshToken = await this.refreshTokenRepository.findOne(token, { relations: ["user"] });

        const isValid =
            !_.isNil(refreshToken) &&
            !_.isNil(refreshToken.user) &&
            (_.isNil(refreshToken.validUntil) || moment().isSameOrBefore(refreshToken.validUntil));

        if (!isValid && !_.isNil(refreshToken)) {
            await this.refreshTokenRepository.delete(refreshToken);
        }

        return !isValid
            ? null
            : {
                  accessToken: await this.generateAuthToken(refreshToken.user),
                  refreshToken
              };
    }
}
