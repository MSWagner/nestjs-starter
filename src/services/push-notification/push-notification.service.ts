import { Inject, Injectable } from "@nestjs/common";
import { Repository } from "typeorm";

import { PushToken } from "../../entities/PushToken.entity";
import { User } from "../../entities/User.entity";

import CONFIG from "../../config";

@Injectable()
export class PushNotificationService {
    constructor(
        @Inject(CONFIG.database.defaultPushTokenRepoName)
        private readonly pushTokenRepository: Repository<PushToken>
    ) {}

    async upsertPushToken(token: string, user: User): Promise<PushToken | null> {
        let pushToken = await PushToken.findOne({ where: { user: user.uid } });
        if (!pushToken) {
            pushToken = new PushToken();
            pushToken.user = user;
        }

        pushToken.token = token;

        try {
            pushToken = await this.pushTokenRepository.save(pushToken);
            return pushToken;
        } catch (error) {
            return null;
        }
    }
}
