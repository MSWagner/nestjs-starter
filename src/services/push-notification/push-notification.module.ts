import { Module } from "@nestjs/common";

import { pushTokenProviders } from "./providers/push-token.provider";
import { PushNotificationService } from "./push-notification.service";
import { UserModule } from "../user/user.module";

@Module({
    imports: [UserModule],
    providers: [PushNotificationService, ...pushTokenProviders],
    exports: [PushNotificationService]
})
export class PushNotificationModule {}
