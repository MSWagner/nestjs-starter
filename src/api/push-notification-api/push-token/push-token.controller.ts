import { Controller, Request, Post, UseGuards, BadRequestException, Body } from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";
import { ApiBody, ApiCreatedResponse, ApiTags } from "@nestjs/swagger";

import { PushTokenDto } from "./_types";
import { SuccessResponse } from "../../auth-api/_types";

import { Permissions } from "../../../services/auth/permissions/permission.decorator";
import { PermissionsGuard } from "../../../services/auth/permissions/permission.guard";
import { PermissionScope } from "../../../entities/Permission.entity";

import { PushNotificationService } from "../../../services/push-notification/push-notification.service";

@ApiTags("auth")
@UseGuards(AuthGuard("bearer"), PermissionsGuard)
@Controller({ path: "push-token", version: "1" })
export class PushTokenController {
    constructor(private readonly pushNotificationService: PushNotificationService) {}

    @Post()
    @Permissions(PermissionScope.User, PermissionScope.Admin)
    @ApiBody({ type: PushTokenDto })
    @ApiCreatedResponse({ description: "The push was successful created or updated.", type: SuccessResponse })
    async handler(@Request() req, @Body() requestDto: PushTokenDto): Promise<SuccessResponse> {
        try {
            const pushToken = await this.pushNotificationService.upsertPushToken(requestDto.token, req.user);
            if (!pushToken) {
                throw new BadRequestException();
            }
        } catch (err) {
            throw new BadRequestException();
        }

        return {
            success: true
        };
    }
}
