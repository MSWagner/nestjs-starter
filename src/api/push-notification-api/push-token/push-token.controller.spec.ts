import request from "supertest";
import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";

import { setupApp } from "../../../appSetup";

import { TestModule } from "../../../services/test/test.module";
import { TestService } from "../../../services/test/test.service";
import { AuthModule } from "../../../services/auth/auth.module";

import { PushTokenController } from "./push-token.controller";
import { PushNotificationModule } from "../../../services/push-notification/push-notification.module";

import * as fixtures from "../../../services/test/fixtures";

describe("PushTokenController", () => {
    let app: INestApplication;
    let testService: TestService;

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            controllers: [PushTokenController],
            imports: [TestModule, PushNotificationModule, AuthModule]
        }).compile();

        testService = module.get<TestService>(TestService);

        app = module.createNestApplication();
        await setupApp(app);
        await app.init();
    });

    beforeEach(async () => {
        await testService.reloadFixtures();
    });

    afterAll((done) => {
        testService.connectionManager.connection.close();
        done();
    });

    it("### PermissionGuard - should return 401 for no access", async () => {
        const body = {
            token: "testToken"
        };

        const response = await request(app.getHttpServer()).post("/api/v1/push-token").send(body).expect(401);
        expect(response.body).toMatchSnapshot("NoPermissionsUserOrAdmin");
    });

    it("### PermissionGuard - should return success with admin permission", async () => {
        const body = {
            token: "testToken"
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/push-token")
            .auth(fixtures.accessTokenUser3Admin.token, { type: "bearer" })
            .send(body)
            .expect(201);
        expect(response.text).toMatchSnapshot("Success");
    });

    it("### PermissionGuard - should return success with user permission", async () => {
        const body = {
            token: "testToken"
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/push-token")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(201);
        expect(response.text).toMatchSnapshot("Success");
    });

    it("### should return 400er for push token conflict", async () => {
        const body = {
            token: fixtures.user3PushToken
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/push-token")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(400);
        expect(response.text).toMatchSnapshot("BadRequestTokenConflict");
    });

    it("### should return 400er for to short push token", async () => {
        const body = {
            token: "12"
        };

        const response = await request(app.getHttpServer())
            .post("/api/v1/push-token")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(400);
        expect(response.text).toMatchSnapshot("BadRequestShortToken");
    });

    it("### should return 400er for missing token", async () => {
        const body = {};

        const response = await request(app.getHttpServer())
            .post("/api/v1/push-token")
            .auth(fixtures.accessToken1.token, { type: "bearer" })
            .send(body)
            .expect(400);
        expect(response.text).toMatchSnapshot("BadRequestMissingToken");
    });
});
