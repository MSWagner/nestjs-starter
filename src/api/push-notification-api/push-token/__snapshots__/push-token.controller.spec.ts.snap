// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PushTokenController ### PermissionGuard - should return 401 for no access: NoPermissionsUserOrAdmin 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`PushTokenController ### PermissionGuard - should return success with admin permission: Success 1`] = `"{"success":true}"`;

exports[`PushTokenController ### PermissionGuard - should return success with user permission: Success 1`] = `"{"success":true}"`;

exports[`PushTokenController ### should return 400er for missing token: BadRequestMissingToken 1`] = `"{"statusCode":400,"message":["token must be longer than or equal to 3 characters","token must be a string"],"error":"Bad Request"}"`;

exports[`PushTokenController ### should return 400er for push token conflict: BadRequestTokenConflict 1`] = `"{"statusCode":400,"message":["token must be longer than or equal to 3 characters","token must be a string"],"error":"Bad Request"}"`;

exports[`PushTokenController ### should return 400er for to short push token: BadRequestShortToken 1`] = `"{"statusCode":400,"message":["token must be longer than or equal to 3 characters"],"error":"Bad Request"}"`;
